### PR TITLE
torch.nn.functional.gumbel_softmax #27078

### PR DESCRIPTION
**Pull request by Naresh1318**
**Comments:**
* Grad check from https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8898 not added
* Double data type as seen in     https://github.com/pytorch/pytorch/blob/848d1ba13a3f77bb6f14656f2d84e2228e37a8ef/test/test_nn.py#L8916 not tested

**Issue:**
https://github.com/pytorch/pytorch/issues/27078

### CODE REVIEW

File: test/cpp/api/functional.cpp [hunk 1]
```diff
@@ -575,6 +575,80 @@ TEST_F(FunctionalTest, LogSigmoid) {
   ASSERT_TRUE(torch::allclose(y, y_exp, 1e-4, 1e-7));
 }
 
+TEST_F(FunctionalTest, GumbelSoftmax) {
+
+    // Test 1: 1D shape, 0 and -1 dim
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: nit: we should indent by 2 spaces for all changes in this PR, because the rest of the codebase does so :D

File: torch/csrc/api/include/torch/nn/options/activation.h [hunk 2]
```diff
@@ -60,6 +60,26 @@ struct TORCH_API LeakyReLUOptions {
...
+
+    /// non-negative scalar temperature
+    TORCH_ARG(double, tau) = 1.0;
+
+    /// returned samples will be discretized as one-hot vectors,
+    /// but will be differentiated as if it is the soft sample in autograd. Default: False
+    TORCH_ARG(bool, hard) = false;
+
+    /// deprecated and has no effect. Default: 1e-10
+    TORCH_ARG(double, eps) = 1e-10;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/activation.h]: We can remove this as we don't carry any legacy from the Python code :D

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 3]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
   return torch::log_sigmoid(input);
 }
 
+inline Tensor gumbel_softmax(const Tensor& logits, const GumbelSoftmaxOptions& options = {}) {
+    if (options.eps() != 1e-10) {
+        std::cerr << "`eps` parameter is deprecated and has no effect." << std::endl;
+    }
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: We can remove this as we don't carry any legacy from the Python code :D

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 4]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
...
+
+    torch::Tensor gumbels = -torch::empty_like(logits).exponential_().log();  // ~Gumbel(0,1)
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: we can write `auto` instead of `torch::Tensor` to make it even simpler :D

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 5]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
...
+    gumbels = (logits + gumbels) / options.tau();  // ~Gumbel(logits, tau)
+    torch::Tensor y_soft = gumbels.softmax(options.dim());
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: same here: `torch::Tensor` -> `auto`

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 6]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
...
+
+    if (options.hard()) {
+        // Straight through.
+        torch::Tensor index = std::get<1>(y_soft.max(options.dim(), true));
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: same here: `torch::Tensor` -> `auto`

Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/activation.h]: it would also be awesome to write:
```suggestion
        auto index = std::get<1>(y_soft.max(options.dim(), /*keepdim=*/true));
```
to make it more clear the purpose of the second argument :D

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 7]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
...
+        auto y_hard = torch::zeros_like(logits).scatter_(options.dim(), index, 1.0);
+        auto ret = y_hard - y_soft.detach() + y_soft;
+        return ret;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: I think we can move the declaration of `ret` before the if statement, and then `return ret` after the if statement, to more closely match the Python version :D

File: torch/csrc/api/include/torch/nn/functional/activation.h [hunk 8]
```diff
@@ -48,6 +48,27 @@ inline Tensor logsigmoid(const Tensor& input) {
...
+    } else {
+        auto ret = y_soft;
+        return ret;
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/functional/activation.h]: ditto here regarding moving `return ret` outside of if statement

File: torch/csrc/api/include/torch/nn/options/activation.h [hunk 9]
```diff
@@ -60,6 +60,26 @@ struct TORCH_API LeakyReLUOptions {
   /// can optionally do the operation in-place. Default: False
   TORCH_ARG(bool, inplace) = false;
 };
+  
+// ============================================================================
+
+/// Options for Gumbel Softmax functional and module.
+struct GumbelSoftmaxOptions {
+    GumbelSoftmaxOptions() {}
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/options/activation.h]: We can remove the default constructor as it's created automatically

File: test/cpp/api/functional.cpp [hunk 10]
```diff
@@ -575,6 +575,80 @@ TEST_F(FunctionalTest, LogSigmoid) {
...
+        ASSERT_EQ(counts.sum().item<int>(), num_draws);
+
+        // Check results are asymptotically as expected
+        auto expected = probs * num_draws;
+        // ~z is approximately N(0,1) for unbiased count
+        auto z = (counts - expected) / (expected * (1 - probs)).sqrt();
+        // A (lazy) approximate 99% two-sided test:
+        // occurs with prob alpha~>=0.01 if unbiased
+        ASSERT_LT(z.abs().max().item<float>(), 2.58);
+    }
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: Thanks so much for the thorough tests! I am wondering whether it's possible to also write a test for `F::gumbel_softmax(logits)`, to test the "no-options" use case. Thanks a lot again!

**Review by yf225 — COMMENTED** [review]
@Naresh1318 Thanks so much for the awesome work! I left some minor comments.

File: test/cpp/api/functional.cpp (hunk 1 above)
Comment (Naresh1318) [reply test/cpp/api/functional.cpp]: I've done that everywhere now

**Review by yf225 — APPROVED** [review]
@Naresh1318 Thanks so much for the awesome work! I will merge it after CI passes :D
